photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala

```diff
@@ -0,0 +1,151 @@
+  *                       time.
+  *                       Recommended:  3 < numCorrections < 10
+  *                       Restriction:  numCorrections > 0
+  * @param isTrackingState Whether to track intermediate states during optimization
+  */
+class LBFGSB (
+               lowerBounds: DenseVector[Double],
+               upperBounds: DenseVector[Double],
+               normalizationContext: Broadcast[NormalizationContext],
+               numCorrections: Int = LBFGSB.DEFAULT_NUM_CORRECTIONS,
+               tolerance: Double = LBFGSB.DEFAULT_TOLERANCE,
+               maxNumIterations: Int = LBFGSB.DEFAULT_MAX_ITER,
+               isTrackingState: Boolean = Optimizer.DEFAULT_TRACKING_STATE)
+  extends Optimizer[DiffFunction](tolerance,
+    maxNumIterations,
+    normalizationContext,
+    isTrackingState){
+
+  /**
+    * Under the hood, this adaptor uses an LBFGSB optimizer from Breeze to optimize functions.
+    * The DiffFunction is modified into a Breeze DiffFunction which the Breeze optimizer can understand.
+    * */
+
+  protected class BreezeOptimization
+  (breezeDiffFunction: BreezeDiffFunction[DenseVector[Double]], initCoefficients: Vector[Double]){
+
+    // The actual workhorse
+    private val breezeStates = BreezeOptimizer.iterations(breezeDiffFunction, initCoefficients.toDenseVector)
+    breezeStates.next()
+
+    def next(state: OptimizerState): OptimizerState = {
+      if (breezeStates.hasNext){
+        val nextBreezeState = breezeStates.next()
+        OptimizerState(
+          nextBreezeState.x,
+          nextBreezeState.adjustedValue,
+          nextBreezeState.adjustedGradient,
+          state.iter+1
+        )
+      } else {
+        state
+      }
+    }
+  }
+
+  protected val BreezeOptimizer = new BreezeLBFGSB(lowerBounds, upperBounds, maxNumIterations, numCorrections, tolerance)
+  @transient
+  protected var breezeOptimization: BreezeOptimization = _
+
+  /**
+    * Initialize breeze optimization engine.
+    *
+    * @param objectiveFunction The objective function to be optimized
+    * @param initState The initial state of the optimizer, prior to starting optimization
+    * @param data The training data
+    */
+
+  protected def init(objectiveFunction: DiffFunction, initState: OptimizerState)(data: objectiveFunction.Data): Unit = {
+    var breezeDiffFunction: BreezeDiffFunction[DenseVector[Double]] = new BreezeDiffFunction[DenseVector[Double]] {
+      override def calculate(x: DenseVector[Double]): (Double, DenseVector[Double]) = {
+        var convertedCoefficients: objectiveFunction.Coefficients = objectiveFunction.convertFromVector(x)
+        var result = objectiveFunction.calculate(data, convertedCoefficients, normalizationContext)
+        (result._1, result._2.toDenseVector)
+      }
+    }
+    breezeOptimization = new BreezeOptimization (breezeDiffFunction, initState.coefficients)
+  }
```

Comment:
Aside from the `breezeOptimizer` and the above function (`init`) everything looks mostly the same as in `LBFGS`. You should take a look at `OWLQN` extends `LBFGS` and do the same for `LBFGSB`.

Reply:
I was trying to just extend LBFGS class instead of extend Optimizer. There are several issues I am facing with:
1. `breezeDiffFunction` need to pass in `DenseVector[Double]` as type otherwise `breezeOptimizer` will report type mismatch when calling iterations function.
2. because of above 1, we need to redefine the nested class `BreezeOptimization` and `init` function. 
3. `var breezeOptimization` cannot be override because it is a var and we need to point it to the new nested class `BreezeOptimization`. 

I don't know what's the good way to extend LBFGS class here. We could talk more on this.